Source/WebCore/rendering/RenderBlockFlow.cpp

```diff
@@ -1193,12 +1214,33 @@ void RenderBlockFlow::determineLogicalLeftPositionForChild(RenderBox& child, App
     // If the child is being centred then the margin calculated to do that has factored in any offset required to
     // avoid floats, so use it if necessary.
 
-    if (style().textAlign() == TextAlignMode::WebKitCenter || child.style().marginStartUsing(&style()).isAuto())
+    if (style().textAlign() == TextAlignMode::WebKitCenter || child.style().marginStartUsing(&style()).isAuto()) {
         newPosition = std::max(newPosition, positionToAvoidFloats + childMarginStart);
-    else if (positionToAvoidFloats > initialStartPosition)
+        setLogicalLeftForChild(child, style().isLeftToRightDirection() ? newPosition : totalAvailableLogicalWidth - newPosition - logicalWidthForChild(child), applyDelta);
+        return;
+    }
+
+    if (positionToAvoidFloats > initialStartPosition)
         newPosition = std::max(newPosition, positionToAvoidFloats);
+    newPosition = computeSelfAlignedLogicalLeftPositionForChild(newPosition);
+    setLogicalLeftForChild(child, newPosition, applyDelta);
```

Comment:
There's an interesting interaction with floats here. Can you explain what's happening, and if it's correct?

Reply:
(e.g. in LTR direction) if a `float: left` exists that is pushing the left edge of the content out we need to use that offset to make sure the content starts after the float, and we use that to compute the start-aligned position of the element

Calling the variable `newPosition` is probably the confusing part, I should really just be modifying the `startPosition` since that's what this is representing

Reply:
the comment explaining this behaviour is a couple lines up but I guess it got decoupled, I'll make this clearer